### ISSUE TITLE
Using cublas_ops with no negative prompt or uncond skip results in black image

### ISSUE THREAD

**Issue by LoganBooker** [Haoming02/sd-webui-forge-classic#19]
Hey Haoming02,

Firstly, thanks for your work on this. Nice to see a lightweight version of Forge without all the bloat. I've essentially moved across to this now after porting a lot of my own improvements from my own hacked version of the old Forge. I'll sort out some pull requests for review once I'm happy with everything.

Now for this issue. I was going to make a pull request but wanted to make sure it wasn't just an issue with my setup. CublasLinear (specifically the call to `cublas_half_matmul` in the forward pass) doesn't seem to like it when b=1. This happens when you have an empty negative prompt or use any of the settings that skip the unconditional. I'm not knowledgeable enough to know if this is an issue with [torch-cublas-hgemm](https://github.com/aredden/torch-cublas-hgemm/tree/master) itself, but here is the workaround I've come up with for Forge Classic:

In `ldm_patched/modules/ops.py`, replace the `cublas_ops` class with:
```
    class cublas_ops(disable_weight_init):
        class Linear(CublasLinear, disable_weight_init.Linear):
            def reset_parameters(self):
                return None

            def forward_ldm_patched_cast_weights(self, input):
                if input is not None and input.shape[0] == 1:
                    dummy = torch.zeros_like(input)
                    input_padded = torch.cat([input, dummy[:1]], dim=0)
                    out_padded = super().forward(input_padded)
                    return out_padded[:1]
                return super().forward(input)

            def forward(self, *args, **kwargs):
                x = kwargs.get('x', args[0] if args else None)
                if x is not None and x.shape[0] == 1:
                    dummy = torch.zeros_like(x)
                    x_padded = torch.cat([x, dummy[:1]], dim=0)
                    if 'x' in kwargs:
                        kwargs['x'] = x_padded
                        out_padded = super().forward(*args, **kwargs)
                    else:
                        new_args = (x_padded,) + args[1:]
                        out_padded = super().forward(*new_args, **kwargs)
                    return out_padded[:1]
                return super().forward(*args, **kwargs)
```

Essentially we pad the tensor so b=2, and then drop the padding on return. Another option would be to fall back to `torch.nn.functional.linear` when b=1 is detected.

I'm running Windows 11 24H2, 4090 Mobile, 561.09 drivers (which are relative old and might be responsible, but NVIDIA's drivers in recent times have been... less than great). It's possible this is a quirk of my OS/hardware/drivers, hence why I'd like confirmation before doing a pull request (or an alternative solution if this isn't the best).

**Comment by Haoming02** [comment]
> doesn't seem to like it when b=1

Hmm, I don't think that is the root cause. Cause in the **Optimization** settings, certain options will skip the negative prompts to speed up the generation. And when I add a `print(input.shape)` to the `forward()`, it does print a batch size of `1` while still generating images correctly. However, I can also repro the black images when generating without a negative prompt...

Will investigate more~

**Comment by LoganBooker** [comment]
Interesting, and thanks for looking into it. It might be something to do with my setup then, as using the skip uncond options in the settings causes the same issue, which was only fixed by padding the batch size or disabling cublas. So not sure what's happening there.

**Comment by Haoming02** [comment]
Oh, mine still generates fine with those options enabled, if I have a negative prompt

~~Interesting indeed~~

I also checked the Comfy repos, and no one has brought up this issue it seems

**Comment by Haoming02** [comment]
In the meantime, you can change the function to
```py
            def forward(self, input: torch.Tensor):
                input = torch.nan_to_num(input, nan=0.0)
                return super().forward(input)
```

Though, it does lower the processing speed a bit...

**Comment by LoganBooker** [comment]
I did a little more investigation on my particular issue, and found this was the culprit (https://github.com/aredden/torch-cublas-hgemm/blob/4e8b688d4037878b67056721b1b7d27551742fed/cublas_ops/__init__.py#L155-L167):
```
    if math.prod(x.shape) == x.shape[-1]:
        out = hgemv_simt(x, weight)
        if use_cublasLt:
            if has_bias:
                out = out + bias.broadcast_to(out.shape)
            if epilogue_str == "RELU":
                out = F.relu(out)
            elif epilogue_str == "GELU":
                out = F.gelu(out)
        if needs_convert:
            out_final = out.type(out_dtype)
        else:
            out_final = out
```
My setup doesn't like `hgemv_simt` kernel -- might be a driver issue possibly. Removing this path and recompiling cublas_ops fixes my issue with no performance impact.

As for the empty negative prompt issue, I was able to solve that by adding to `sd_samplers_cfg_denoiser.py` around line 154:
```
        if not torch.any(uncond["crossattn"]):
            cond_scale = 1.0
```
Feels a bit hacky, I'm sure there's a more elegant solution.

**Comment by LoganBooker** [comment]
OK, I've got this figured out, and it's all to do with using SageAttention 2. Uninstalling 2.1.1 and installing 1.0.6 fixed the empty prompt issue, as did using `--disable-sage` (which is how I discovered it was Sage in the first place).

I was under the impression I had tested without Sage before making the issue by removing `--sage` from the command line, but you need to explicitly specify `--disable-sage` to actually disable it.

Apologies for the runaround, and do appreciate the time you spent looking into this.

**EDIT:** Still have the issue with cublas_ops and b=1 though, but I have an effective fix for that now.
